timber/src/test/java/timber/log/TimberTest.kt

```diff
@@ -175,8 +175,8 @@ class TimberTest {
     }.run()
 
     assertLog()
-        .hasDebugMessage("TimberTest\$debugTreeTag", "Hello, world!")
-        .hasDebugMessage("TimberTest\$debugTreeTag", "Hello, world!")
+        .hasDebugMessage("TimberTest", "Hello, world!")
+        .hasDebugMessage("TimberTest", "Hello, world!")
```

Comment:
this doesn't seem right.  I'll dig in more tomorrow, but here's a before/after of a debugger snapshot after merging https://github.com/JakeWharton/timber/pull/417

Before:
![Screen Shot 2021-08-03 at 1 26 54 AM](https://user-images.githubusercontent.com/1868149/127963388-fcaa5c0a-5916-4de9-a209-7443611a2942.png)

After:
![Screen Shot 2021-08-03 at 1 33 17 AM](https://user-images.githubusercontent.com/1868149/127963398-1df06682-d60f-4695-9c4f-ec2c36786b23.png)

Reply:
Interesting. I only tested this on top of #417 so I assumed this test was just broken. Maybe it depends on the version of the jvm?

Is it possible the junit version bump did something here?

Reply:
Ok, looks like the Kotlin bump to 1.5.*; I _think_ due to a change in how SAM-interfaces and lambdas are generated per: https://kotlinlang.org/docs/whatsnew15.html#sam-adapters-via-invokedynamic

In the failing test, we have the following snippet:
```
    Runnable {
      Timber.d("Hello, world!")

      Runnable { Timber.d("Hello, world!") }.run()
    }.run()
```

with the following top of stack trace.
```
0 = {StackTraceElement@4236} "timber.log.Timber$DebugTree.getTag$timber_debug(Timber.kt:205)"
1 = {StackTraceElement@4237} "timber.log.Timber$Tree.prepareLog(Timber.kt:148)"
2 = {StackTraceElement@4238} "timber.log.Timber$Tree.d(Timber.kt:51)"
3 = {StackTraceElement@4239} "timber.log.Timber$Forest.d(Timber.kt:292)"
```

Then I compared the difference when I changed `Runnable {...}` to its equivalent anonymous class `object : Runnable {...}`

When outer is anonymous class
```
4 = {StackTraceElement@4240} "timber.log.TimberTest$debugTreeTagGenerationStripsAnonymousClassMarker$1.run(TimberTest.kt:173)"
```

...and inner is anonymous class
```
4 = {StackTraceElement@4697} "timber.log.TimberTest$debugTreeTagGenerationStripsAnonymousClassMarker$1$run$1.run(TimberTest.kt:177)"
```
...and inner is SAM lambda
```
4 = {StackTraceElement@4335} "timber.log.TimberTest$debugTreeTagGenerationStripsAnonymousClassMarker$1.run$lambda-0(TimberTest.kt:175)"
```

When outer is SAM lambda
```
4 = {StackTraceElement@4227} "timber.log.TimberTest.debugTreeTagGenerationStripsAnonymousClassMarker$lambda-4(TimberTest.kt:172)"
```

...and inner is anonymous class
```
4 = {StackTraceElement@4328} "timber.log.TimberTest$debugTreeTagGenerationStripsAnonymousClassMarker$1$1.run(TimberTest.kt:176)"
```

...and inner is SAM lambda
```
4 = {StackTraceElement@4328} "timber.log.TimberTest.debugTreeTagGenerationStripsAnonymousClassMarker$lambda-4$lambda-3(TimberTest.kt:174)"
```

Reply:
The outer-SAM inner-anonymous case is interesting. 

So does the library need to be updated to account for `LambdaMetafactory`? Is `lambda-[0-9]+` spec or is it implementation specific?

Reply:
I split this into 5 tests:

1. Only anonymous classes
2. Only SAM lambdas
3. Only lambdas
4. Outer anonymous class inner SAM lambda
5. Outer SAM lambda inner anonymous class